Source/mpq/mpq_writer.cpp

```diff
@@ -218,111 +234,102 @@ _BLOCKENTRY *MpqWriter::NewBlock(int *blockIndex)
 
 void MpqWriter::AllocBlock(uint32_t blockOffset, uint32_t blockSize)
 {
-	_BLOCKENTRY *block;
-	int i;
-
+	MpqBlockEntry *block;
+expanded:

```

Comment:
What does this do?

Reply:
It's the label for the gotos. I guess the function always returns after the recursion so there's no need to allocate the extra stack space if we simulate the tail recursion by instead jumping back to the beginning of the function.

Wouldn't this still be pretty easy to implement by surrounding the for loop in a while loop? Perhaps we should avoid the gotos here.

Reply:
I'm not sure how to implement it with a while loop without an extra boolean. It seemed pretty confusing when I tried it that way. `goto` is sometimes (but rarely so) the simplest solution.

Reply:
You would need another bool for the while loop, but it would essentially take the place of the label. Conceptually, it's already pretty difficult to understand what's going on in this loop, and I don't think the while loop really adds any confusion compared to the goto. I'd argue that the curly braces and nesting should help to see the structure of the nested loop here, even if it doesn't really help to clarify the intent.

```c++
	MpqBlockEntry *block;
	bool expanded = true;
	while (expanded) {
		expanded = false;
		block = blockTable_;
		for (unsigned i = BlockEntriesCount; i-- != 0; ++block) {
			// Expand to adjacent blocks.
			if (!IsAllocatedUnusedBlock(block))
				continue;
			if (block->offset + block->packedSize == blockOffset) {
				blockOffset = block->offset;
				blockSize += block->packedSize;
				memset(block, 0, sizeof(MpqBlockEntry));
				expanded = true;
				break;
			}
			if (blockOffset + blockSize == block->offset) {
				blockSize += block->packedSize;
				memset(block, 0, sizeof(MpqBlockEntry));
				expanded = true;
				break;
			}
		}
	}
```

Reply:
> I'd argue that the curly braces and nesting should help to see the structure of the nested loop

I much agree with this. Goto is easy to miss and break.

Other then that I'm happy with this PR at this point 👍

Reply:
I find it a lot more readable with `goto` but I guess I'm more used to `goto` from reading lots of lower-level performance-optimized C code. Done.